### ISSUE TITLE
scala AbstractKeys cannot be accessed

### ISSUE THREAD

**Issue by cintoros** [jOOQ/jOOQ#6875]
### Expected behavior and actual behavior:
building the generated classes results in:
```
Error:(50, 78) method createIndex in object AbstractKeys cannot be accessed in object org.jooq.impl.AbstractKeys
 Access to protected method createIndex not permitted because
 prefix type org.jooq.impl.AbstractKeys.type does not conform to
 object Indexes0 in object Indexes where the access takes place
    val FETCHFLOW_DOMAIN_FETCHFLOW_DOMAIN_DOMAIN_HASH : Index = AbstractKeys.createIndex("fetchflow_domain_domain_hash", FetchflowDomain.FETCHFLOW_DOMAIN, Array[OrderField [_] ](FetchflowDomain.FETCHFLOW_DOMAIN.DOMAIN_HASH), false)
```

### Steps to reproduce the problem:

we use following scala code to generate our jooq schemas

```
    val configuration = new Configuration()
      .withJdbc(new Jdbc()
        .withDriver("org.mariadb.jdbc.Driver")
        .withUrl(s"jdbc:mysql://localhost/$schema")
        .withUser("root")
        .withPassword(""))
      .withGenerator(new Generator()
        .withName("org.jooq.util.ScalaGenerator")
        .withDatabase(new Database()
          .withName("org.jooq.util.mariadb.MariaDBDatabase")
          .withInputSchema(schema))
        .withTarget(new Target()
          .withPackageName(packageName)
          .withDirectory(directory))
      )

    GenerationTool.generate(configuration)
```

### Versions:

- jOOQ: 3.10.1
- Java: 1.8
- Database (include vendor): mariadb
- JDBC Driver (include name if inofficial driver): mariadb

**Comment by lukaseder** [comment]
Thanks for your report. What Scala version are you using?

**Comment by cintoros** [comment]
Scala Version 2.12.4

**Comment by lukaseder** [comment]
Thanks. I'll look into this shortly.

For the record, there seems to be a similar bug report on the mailing list, here:
https://groups.google.com/forum/#!topic/jooq-user/UihBK0pR-YU

**Comment by lukaseder** [comment]
There has definitely been a regression caused by the Scala version 2.12.4. I can reproduce many similar errors, which weren't there in 2.12.3. Will continue investigating

**Comment by er1c** [comment]
I can confirm this issue too, I'll see if i can come up with a good fix

**Comment by er1c** [comment]
Related: https://github.com/scala/bug/issues/10597

**Comment by lukaseder** [comment]
Thanks for finding those links, @er1c. I'll be observing those issues. As I thought, this should be fixed in the scalac. While we could open up `AbstractKeys` methods to be public, I'd rather not to, if the scalac issue can be fixed in a timely fashion... From the release notes' release cadence, it looks like that could happen quite soon: https://github.com/scala/scala/releases

**Comment by lukaseder** [comment]
A workaround is to use reflection in generated code. You could post process generated code by replacing calls to `AbstractKeys` methods as follows:

**Search**

```java
AbstractKeys.createUniqueKey(TAuthor.T_AUTHOR, "PK_T_AUTHOR", TAuthor.T_AUTHOR.ID)
```

**Replace**

```java
org.jooq.tools.reflect.Reflect.on(classOf[AbstractKeys]).call("createUniqueKey", TAuthor.T_AUTHOR, "PK_T_AUTHOR", TAuthor.T_AUTHOR.ID).get()
```

This seems to work for me. A regex that could work here is:

```
replaceAll(
  "AbstractKeys\\.(\\w+)\\((.*?)\n", 
  "org.jooq.tools.reflect.Reflect.on(classOf[AbstractKeys]).call(\"$1\", $2.get()\n"
);
```

**Comment by lukaseder** [comment]
OK, this fix in the scala compiler might not be published too soon, I guess. We'll fix this on the jOOQ side by creating a new public type `Internal`, with all public static methods and avoid protected static in the future: #7158

**Comment by lukaseder** [comment]
This week, we'll publish jOOQ 3.10.5 with a fix for #7158.
